Makefile.defs

```diff
@@ -33,70 +41,65 @@ ifneq ($(wildcard $(dir $(lastword $(MAKEFILE_LIST)))/.git),)
+GO_BUILD_LDFLAGS += -X "github.com/cilium/cilium/pkg/version.Version=$(FULL_BUILD_VERSION)"
 
 ifeq ($(NOSTRIP),)
 	# Note: these options will not remove annotations needed for stack
 	# traces, so panic backtraces will still be readable.
 	#
 	# -w: Omit the DWARF symbol table.
 	# -s: Omit the symbol table and debug information.
-	GOLDFLAGS += -s -w
+	GO_BUILD_LDFLAGS += -s -w
 endif
 
 CILIUM_ENVOY_SHA=$(shell grep -o "FROM.*cilium/cilium-envoy:[0-9a-fA-F]*" $(ROOT_DIR)/Dockerfile | cut -d : -f 2)
-GOLDFLAGS += -X "github.com/cilium/cilium/pkg/envoy.RequiredEnvoyVersionSHA=$(CILIUM_ENVOY_SHA)"
+GO_BUILD_LDFLAGS += -X "github.com/cilium/cilium/pkg/envoy.RequiredEnvoyVersionSHA=$(CILIUM_ENVOY_SHA)"
 
 BPF_FILES_EVAL := $(shell git ls-files $(ROOT_DIR)/bpf/ | grep -v .gitignore | tr "\n" ' ')
 BPF_FILES ?= $(BPF_FILES_EVAL)
 BPF_SRCFILES := $(subst ../,,$(BPF_FILES))
 
 CILIUM_DATAPATH_SHA=$(shell cat $(BPF_FILES) | sha1sum | awk '{print $$1}')
-GOLDFLAGS += -X "github.com/cilium/cilium/pkg/datapath/loader.DatapathSHA=$(CILIUM_DATAPATH_SHA)"
-
-# Set DOCKER_IMAGE_TAG with "latest" by default
-ifeq ($(DOCKER_IMAGE_TAG),)
-    DOCKER_IMAGE_TAG="latest"
-endif
+GO_BUILD_LDFLAGS += -X "github.com/cilium/cilium/pkg/datapath/loader.DatapathSHA=$(CILIUM_DATAPATH_SHA)"
 
 # Set -mod=vendor if running >= go 1.13 or if GO111MODULE is set.
 # A go build is being executed with go modules if:
 # * The go command is invoked with GO111MODULE=on environment variable set.
 # * The go command is invoked in a directory outside of the $GOPATH/src tree
 #   and the environment variable GO111MODULE unset (or explicitly set to 'auto').
 ifeq ($(GO111MODULE),on)
-    GOFLAGS ?= -mod=vendor
-    GOCLEAN ?= -mod=vendor
+    GO_BUILD_FLAGS += -mod=vendor
+    GO_TEST_FLAGS += -mod=vendor
+    GO_CLEAN_FLAGS += -mod=vendor
 endif
+
 # Compile with '-mod=vendor' if go >= 1.13
 GO_MAJOR_VERSION_GE_1 := $(shell expr `$(GO) version | grep -E 'go[0-9]{1}+' -o | sed 's/go//g'` \>= 1)
 ifeq ($(GO_MAJOR_VERSION_GE_1),1)
     GO_MINOR_VERSION_GE_13 := $(shell expr `$(GO) version | grep -E 'go[^ ]+' -o | sed 's/go1.//g'` \>= 13)
     ifeq ($(GO_MINOR_VERSION_GE_13),1)
-        GOFLAGS ?= -mod=vendor
-        GOCLEAN ?= -mod=vendor
+        GO_BUILD_FLAGS += -mod=vendor
+        GO_TEST_FLAGS += -mod=vendor
+        GO_CLEAN_FLAGS += -mod=vendor
     endif
 endif
 
-GOBUILD = $(GOFLAGS) -ldflags '$(GOLDFLAGS)' $(EXTRA_GOBUILD_FLAGS)
-
-# Uncomment to enable race detection
-#GOBUILD += -race
-
-# Uncomment to enable deadlock detection
-#GOBUILD += -tags lockdebug
+ifneq ($(RACE),)
+    GO_BUILD_FLAGS += -race
+    GO_TEST_FLAGS += -race
+endif
 
 ifneq ($(LOCKDEBUG),)
-    GOBUILD += -tags lockdebug
+    GO_BUILD_FLAGS += -tags lockdebug
 endif
 
-# Container engine
-export CONTAINER_ENGINE?=docker
-CONTAINER_ENGINE_FULL=$(QUIET)$(CONTAINER_ENGINE)
+GO_BUILD_FLAGS += -ldflags '$(GO_BUILD_LDFLAGS) $(EXTRA_GO_BUILD_LDFLAGS)' $(EXTRA_GO_BUILD_FLAGS)
```

Comment:
When are these 'variables' dereferenced exactly? Is it at dynamically at the time that the target is run? Mainly concerned that the proxylib invocation that uses extra flags no longer explicitly passes the flags.

Reply:
(Also, RIP golden flags, those shiny flags always gave me a chuckle when I read them)

Reply:
`+=` is non immediate is this case, perhaps I should make it more explicit with a declaration and a comment to clarify.

https://www.gnu.org/software/make/manual/html_node/Reading-Makefiles.html